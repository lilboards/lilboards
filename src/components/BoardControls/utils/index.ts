export * from './sortByLikes';
export * from './timer';
export * from './transform';
