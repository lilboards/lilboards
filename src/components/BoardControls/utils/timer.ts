import { MINUTE_IN_SECONDS, SECOND_IN_MILLISECONDS } from '../constants';

/**
 * Formats milliseconds remaining to `mm:ss`.
 */
export function formatTimeRemaining(milliseconds: number) {
  let seconds = Math.round(milliseconds / SECOND_IN_MILLISECONDS);
  const minutes = Math.floor(seconds / MINUTE_IN_SECONDS);
  seconds = seconds % MINUTE_IN_SECONDS;
  return `${minutes || '00'}:${String(seconds).padStart(2, '0')}`;
}

/**
 * Converts minutes to milliseconds.
 */
export const minutesToMilliseconds = (minutes: number) =>
  minutes * MINUTE_IN_SECONDS * SECOND_IN_MILLISECONDS;
