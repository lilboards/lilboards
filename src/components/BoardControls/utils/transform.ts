import type { Columns, Items } from 'src/types';

const NEWLINE = '\n';

/**
 * Transforms board data to a multidimensional array.
 *
 * @returns - Multidimensional array.
 */
export function transformToRows(columns: Columns, items: Items) {
  const columnValues = Object.values(columns);
  const columnsLength = columnValues.length;

  if (!columnsLength) {
    return [];
  }

  const rows: string[][] = [];
  const headers = columnValues.map(
    (column, index) => column.name || `Column ${index + 1}`,
  );
  rows.push(headers);

  columnValues.forEach(({ itemIds }, columnIndex) => {
    if (!itemIds || !itemIds.length) {
      return;
    }

    itemIds.forEach((itemId, rowIndex) => {
      rowIndex++; // don't override headers
      rows[rowIndex] = rows[rowIndex] || [...Array(columnsLength)];
      rows[rowIndex][columnIndex] = items[itemId].text;
    });
  });

  return rows;
}

/**
 * Transforms board data to CSV.
 *
 * @param columns - Columns.
 * @param items - Items.
 * @returns - CSV.
 */
export function transformToCSV(columns: Columns, items: Items): string {
  const rows = transformToRows(columns, items);

  if (!rows.length) {
    return '';
  }

  let csv = '';
  const headers = rows.shift()!;
  csv += headers.map((header) => JSON.stringify(header)).join(',');
  csv += NEWLINE;

  rows.forEach((row) => {
    csv += row.map((column) => JSON.stringify(column)).join(',');
    csv += NEWLINE;
  });

  return csv.trim();
}

/**
 * Transforms board data to Markdown.
 *
 * @param columns - Columns.
 * @param items - Items.
 * @returns - Markdown.
 */
export function transformToMarkdown(columns: Columns, items: Items): string {
  const rows = transformToRows(columns, items);

  if (!rows.length) {
    return '';
  }

  let markdown = '';
  const headers = rows.shift()!;
  markdown += `| ${headers.join(' | ')} |`;
  markdown += NEWLINE;
  markdown += `|${headers.map(() => ' --- |').join('')}`;
  markdown += NEWLINE;

  rows.forEach((row) => {
    markdown += `| ${row
      .map((text) =>
        typeof text === 'string' ? text.replace(/\n/g, '<br>') : text,
      )
      .join(' | ')} |`;
    markdown += NEWLINE;
  });

  return markdown.trim();
}
