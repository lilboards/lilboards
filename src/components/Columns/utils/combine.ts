import type { DropResult } from '@hello-pangea/dnd';
import type { Items, LikesItems } from 'src/types';

const separator = '\n\n---\n\n';

/**
 * @see {@link https://github.com/hello-pangea/dnd/blob/main/docs/guides/combining.md}
 */
export function combine(result: DropResult, items: Items, likes: LikesItems) {
  const sourceItemId = result.draggableId;
  const sourceColumnId = result.source.droppableId;
  const sourceText = items[sourceItemId].text.trim();

  const targetItemId = result.combine!.draggableId;
  const targetText = items[targetItemId].text.trim();

  const combinedText = [targetText, sourceText].join(separator);
  const combinedLikes = {
    ...likes[sourceItemId],
    ...likes[targetItemId],
  };

  return {
    remove: {
      itemId: sourceItemId,
      columnId: sourceColumnId,
    },
    update: {
      itemId: targetItemId,
      likes: combinedLikes,
      text: combinedText,
    },
  };
}
