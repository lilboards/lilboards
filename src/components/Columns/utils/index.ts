export * from './combine';
export * from './reorder';
