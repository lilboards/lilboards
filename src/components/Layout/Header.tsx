import AppBar from '@material-ui/core/AppBar';
import Box from '@material-ui/core/Box';
import Button from '@material-ui/core/Button';
import Container from '@material-ui/core/Container';
import Link from '@material-ui/core/Link';
import Toolbar from '@material-ui/core/Toolbar';

import { Link as RouterLink, useLocation } from '@reach/router';

import { REDIRECT_TO } from '../../constants';
import { useSelector } from '../../hooks';

export default function Header() {
  const isLoggedIn = useSelector((state) => Boolean(state.user.email));
  const location = useLocation();

  return (
    <AppBar position="static">
      <Container>
        <Toolbar disableGutters>
          <Box flexGrow={1}>
            <Link color="inherit" component={RouterLink} to="/" variant="h6">
              Lilboards
            </Link>
          </Box>

          {isLoggedIn && (
            <Box marginRight={1}>
              <Button color="inherit" component={RouterLink} to="/boards">
                Boards
              </Button>
            </Box>
          )}

          <Button
            color="inherit"
            component={RouterLink}
            state={{ [REDIRECT_TO]: location.pathname }}
            to={isLoggedIn ? '/logout' : '/login'}
            variant="outlined"
          >
            {isLoggedIn ? 'Logout' : 'Login'}
          </Button>
        </Toolbar>
      </Container>
    </AppBar>
  );
}
