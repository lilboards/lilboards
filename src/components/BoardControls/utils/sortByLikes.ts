import type { ColumnItemIds, Columns, Likes } from 'src/types';
import { cloneArray, countObject } from 'src/utils';

export function sortByLikes(columns: Columns, likes: Likes): ColumnItemIds {
  const columnItemIds: ColumnItemIds = {};

  for (const columnId in columns) {
    const { itemIds } = columns[columnId];

    columnItemIds[columnId] = cloneArray(itemIds).sort((itemId1, itemId2) => {
      const likes1 = countObject(likes.items[itemId1]);
      const likes2 = countObject(likes.items[itemId2]);
      return likes2 - likes1;
    });
  }

  return columnItemIds;
}
