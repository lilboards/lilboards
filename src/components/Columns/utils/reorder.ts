import type { DraggableLocation } from '@hello-pangea/dnd';
import { DatabaseKey } from 'src/constants';
import type { Columns } from 'src/types';
import { cloneArray, reorderArray } from 'src/utils';

export function reorder(
  source: DraggableLocation,
  destination: DraggableLocation,
  columns: Columns,
) {
  const sourceColumnId = source.droppableId;
  const sourceColumn = columns[sourceColumnId] || {};
  const sourceItemIds = cloneArray(sourceColumn[DatabaseKey.itemIds]);

  const destinationColumnId = destination.droppableId;
  const destinationColumn = columns[destinationColumnId] || {};
  const destinationItemIds = cloneArray(destinationColumn[DatabaseKey.itemIds]);

  // moving item to the same column
  if (sourceColumnId === destinationColumnId) {
    return {
      [sourceColumnId]: reorderArray(
        sourceItemIds,
        source.index,
        destination.index,
      ),
    };
    // moving item to a different column
  } else {
    // insert item into the next column and remove item from the original column
    const targetItemId = sourceItemIds[source.index];
    destinationItemIds.splice(destination.index, 0, targetItemId);
    sourceItemIds.splice(source.index, 1);

    return {
      [sourceColumnId]: sourceItemIds,
      [destinationColumnId]: destinationItemIds,
    };
  }
}
