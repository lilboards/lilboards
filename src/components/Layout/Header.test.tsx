import { screen } from '@testing-library/react';

import { renderWithContext, updateStore } from '../../utils/test';
import Header from './Header';

it('renders header', () => {
  renderWithContext(<Header />);
  expect(screen.getByRole('banner')).toHaveTextContent('Lilboards');
});

it('renders heading link', () => {
  renderWithContext(<Header />);
  expect(screen.getByRole('link', { name: 'Lilboards' })).toHaveAttribute(
    'href',
    '/'
  );
});

it('renders GitHub link', () => {
  renderWithContext(<Header />);
  expect(screen.getByLabelText('Open GitHub repository')).toHaveAttribute(
    'href',
    'https://b.remarkabl.org/lilboards'
  );
});

describe('when not logged in', () => {
  it('renders login link', () => {
    updateStore.withUser(false);
    renderWithContext(<Header />);
    expect(screen.getByRole('link', { name: 'Login' })).toHaveAttribute(
      'href',
      '/login'
    );
  });

  it('does not render boards link', () => {
    updateStore.withUser(false);
    renderWithContext(<Header />);
    expect(
      screen.queryByRole('link', { name: 'Boards' })
    ).not.toBeInTheDocument();
  });
});

describe('when logged in', () => {
  it('renders logout link', () => {
    updateStore.withUser();
    renderWithContext(<Header />);
    expect(screen.getByRole('link', { name: 'Logout' })).toHaveAttribute(
      'href',
      '/logout'
    );
  });

  it('renders boards link', () => {
    updateStore.withUser();
    renderWithContext(<Header />);
    expect(screen.getByRole('link', { name: 'Boards' })).toHaveAttribute(
      'href',
      '/boards'
    );
  });
});
