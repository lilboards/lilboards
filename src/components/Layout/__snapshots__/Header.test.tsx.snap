// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <header
    class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular"
      >
        <div
          class="MuiBox-root MuiBox-root-1"
        >
          <a
            aria-current="page"
            class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-h6 MuiTypography-colorInherit"
            href="/"
          >
            Lilboards
          </a>
        </div>
        <a
          aria-disabled="false"
          aria-label="Open GitHub repository"
          class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiTypography-colorPrimary"
          href="https://github.com/lilboards/lilboards"
          rel="noopener noreferrer"
          role="button"
          tabindex="0"
          target="_blank"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 .3a12 12 0 0 0-3.8 23.4c.6.1.8-.3.8-.6v-2c-3.3.7-4-1.6-4-1.6-.6-1.4-1.4-1.8-1.4-1.8-1-.7.1-.7.1-.7 1.2 0 1.9 1.2 1.9 1.2 1 1.8 2.8 1.3 3.5 1 0-.8.4-1.3.7-1.6-2.7-.3-5.5-1.3-5.5-6 0-1.2.5-2.3 1.3-3.1-.2-.4-.6-1.6 0-3.2 0 0 1-.3 3.4 1.2a11.5 11.5 0 0 1 6 0c2.3-1.5 3.3-1.2 3.3-1.2.6 1.6.2 2.8 0 3.2.9.8 1.3 1.9 1.3 3.2 0 4.6-2.8 5.6-5.5 5.9.5.4.9 1 .9 2.2v3.3c0 .3.1.7.8.6A12 12 0 0 0 12 .3"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </a>
        <div
          class="MuiBox-root MuiBox-root-2"
        >
          <a
            aria-disabled="false"
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-colorInherit"
            href="/login"
            role="button"
            tabindex="0"
          >
            <span
              class="MuiButton-label"
            >
              Login
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
        </div>
      </div>
    </div>
  </header>
</div>
`;
